// tslint/build/rules/noDuplicateVariableRule.js
/*
 * Copyright 2013 Palantir Technologies, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/
var __extends = this.__extends || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var Rule = (function (_super) {
    __extends(Rule, _super);
    function Rule() {
        _super.apply(this, arguments);
    }
    Rule.prototype.apply = function (sourceFile) {
        return this.applyWithWalker(new NoDuplicateVariableWalker(sourceFile, this.getOptions()));
    };
    Rule.FAILURE_STRING = "duplicate variable: '";
    return Rule;
})(Lint.Rules.AbstractRule);
exports.Rule = Rule;
var NoDuplicateVariableWalker = (function (_super) {
    __extends(NoDuplicateVariableWalker, _super);
    function NoDuplicateVariableWalker() {
        _super.apply(this, arguments);
    }
    NoDuplicateVariableWalker.prototype.createScope = function () {
        return new ScopeInfo();
    };
    NoDuplicateVariableWalker.prototype.visitParameterDeclaration = function (node) {
        var propertyName = node.name;
        var variableName = propertyName.text;
        var currentScope = this.getCurrentScope();
        if (currentScope.varNames.indexOf(variableName) >= 0) {
            this.addFailureOnIdentifier(node.name);
        }
        else {
            currentScope.varNames.push(variableName);
        }
        _super.prototype.visitParameterDeclaration.call(this, node);
    };
    NoDuplicateVariableWalker.prototype.visitTypeLiteral = function (node) {
    };
    NoDuplicateVariableWalker.prototype.visitMethodSignature = function (node) {
    };
    NoDuplicateVariableWalker.prototype.visitFunctionType = function (node) {
    };
    NoDuplicateVariableWalker.prototype.visitCatchClause = function (node) {
        this.visitBlock(node.block);
    };
    NoDuplicateVariableWalker.prototype.visitVariableDeclaration = function (node) {
        var propertyName = node.name;
        var variableName = propertyName.text;
        var currentScope = this.getCurrentScope();
        var declarationIsLet = (Math.floor(node.parent.flags / 4096) % 2) === 1;
        if (currentScope.varNames.indexOf(variableName) >= 0) {
            this.addFailureOnIdentifier(propertyName);
        }
        else if (!declarationIsLet) {
            if (currentScope.letNames.indexOf(variableName) >= 0) {
                this.addFailureOnIdentifier(propertyName);
            }
            else {
                currentScope.varNames.push(variableName);
            }
        }
        else {
            currentScope.letNames.push(variableName);
        }
        _super.prototype.visitVariableDeclaration.call(this, node);
    };
    NoDuplicateVariableWalker.prototype.addFailureOnIdentifier = function (ident) {
        var failureString = Rule.FAILURE_STRING + ident.text + "'";
        this.addFailure(this.createFailure(ident.getStart(), ident.getWidth(), failureString));
    };
    return NoDuplicateVariableWalker;
})(Lint.ScopeAwareRuleWalker);
var ScopeInfo = (function () {
    function ScopeInfo() {
        this.varNames = [];
        this.letNames = [];
    }
    return ScopeInfo;
})();
